loopy/transform/iname.py

```diff
@@ -1121,26 +1123,49 @@ def has_schedulable_iname_nesting(kernel):
 # {{{ rename_inames
 
 @for_each_kernel
-def rename_iname(kernel, old_iname, new_iname, existing_ok=False, within=None):
+def rename_inames(kernel, old_inames, new_iname, existing_ok=False, within=None):
     """
+    :arg old_inames: A collection of inames that must be renamed to **new_iname**.
     :arg within: a stack match as understood by
         :func:`loopy.match.parse_stack_match`.
     :arg existing_ok: execute even if *new_iname* already exists
     """
+    from collections.abc import Collection
+    if (isinstance(old_inames, str)
+            or not isinstance(old_inames, Collection)):
+        raise LoopyError("'old_inames' must be a collection of strings, "
+                         f"got '{type(old_inames)}'.")
+
+    if any((len(insn.within_inames & frozenset(old_inames)) > 1)
+           for insn in kernel.instructions):
+        raise LoopyError("old_inames contains nested inames"
+                         " -- renaming is illegal.")
+
+    # sort to have deterministic implementation.
+    old_inames = sorted(old_inames)
 
     var_name_gen = kernel.get_var_name_generator()
 
     # FIXME: Distinguish existing iname vs. existing other variable
     does_exist = var_name_gen.is_name_conflicting(new_iname)
 
-    if old_iname not in kernel.all_inames():
-        raise LoopyError("old iname '%s' does not exist" % old_iname)
+    if not (frozenset(old_inames) <= kernel.all_inames()):
+        raise LoopyError(f"old inames {frozenset(old_inames) - kernel.all_inames()}"
+                         " do not exist.")
 
     if does_exist and not existing_ok:
-        raise LoopyError("iname '%s' conflicts with an existing identifier"
-                "--cannot rename" % new_iname)
+        raise LoopyError(f"iname '{new_iname}' conflicts with an existing identifier"
+                         " --cannot rename")
+
+    if not does_exist:
+        kernel = duplicate_inames(
+                kernel, old_inames[0], within=within, new_inames=[new_iname])
+        kernel = remove_unused_inames(kernel, old_inames[0])
+        old_inames = old_inames[1:]
```

Comment:
* Won't this damage the substitution logic below? (Because the `old_inames[0]` -> `new_iname` substitution in the code will never happen? (Answer by comment if possible.)
* Maybe make sure the tests cover both the `does_exist` and `not does_exist` cases.

Reply:
> Won't this damage the substitution logic below?

It shouldn't. Note that we've already renamed `old_iname[0]` to `new_iname` before removing it from the list.